### PR TITLE
feat(judges): wire LLMJudge to Anthropic / OpenAI-compat / Ollama

### PR DESCRIPTION
**Pull request by rsavitt**
## Summary

Replaces arm B's `NotImplementedError` stub with real one-shot LLM calls so the calibration study can actually run end-to-end with three independent judges (Claude / GPT-4o-mini / Llama).

**Stacked on #485 (arm B).** Base on this PR is `feat/calibration-arm-b`.

### What's wired

- **`swarm/judges/llm_call.py`** — synchronous one-shot dispatch:
  - `call_anthropic` (Claude)
  - `call_openai_compatible` (OpenAI / OpenRouter / Groq / Together / DeepSeek — anything that speaks the chat API)
  - `call_ollama` (local Llama-family; no API key)
  - `call_with_retries` — generic exponential-backoff wrapper with an **injectable `sleep`** so tests don't burn wall-clock waiting.
  - `_resolve_api_key` — explicit > `ANTHROPIC_API_KEY` / `OPENAI_API_KEY` / ... > none.

  Deliberately separate from `swarm/agents/llm_agent.py`: a one-shot judge call has no business pulling in reputation, memory, action parsing, or async event loops.

- **`swarm/judges/judge.py`** — `LLMJudge.score()` now real:
  - `_dispatch` routes to the right provider call based on `provider` field.
  - `_extract_score` tries `json.loads` → markdown-fenced ` ```json {...} ``` ` → first `{...}` containing `"score"`. Out-of-range clamped; garbage raises `ValueError` tagged with the interaction id so failures point at the offending item instead of producing fake defaults.
  - `caller` field lets tests inject a fake without network.

- **`experiments/calibration_judge.py`** — `JUDGE_SPECS` map dispatches `claude` / `gpt4o_mini` / `llama` to real providers; override models via `JUDGE_MODEL_<NAME>` env var.

### Tests (32 total, all pass; ruff + mypy clean)

15 new for `LLMJudge`: 5 JSON-parse regimes (clean, fenced, narrated, clamped, garbage), retry-then-succeed with attempt counting, orthogonality enforced in prompt build, dispatch rejects unknown provider. 8 new in `tests/test_llm_call.py` for the standalone helpers (api-key resolution including Ollama keyless, retry semantics: no-retry-on-success, retry-then-succeed, exhausts-and-raises, backoff doubles).

### Smoke

Mock still works:
```
$ python -m experiments.calibration_judge --judges mock --per-bin 5 --seed 1
pool=280, sampled=25, judges=['mock']
```

Real run (requires `ANTHROPIC_API_KEY`, `OPENAI_API_KEY`, and Ollama running):
```
python -m experiments.calibration_judge --judges claude gpt4o_mini llama --per-bin 50 --seed 42
```

Not run in this PR — needs credentials and budget.

## Beads

Unblocks the rest of arm B's actual data collection. Arms C and D can now consume real `judge_scores.csv` instead of the mock-only smoke output. Adaptive arm 0 (`u4av`) gets a real anchor as soon as arm D's joined CSV runs against this.

## Test plan

- [ ] Confirm the JSON-extraction strategies cover real model behavior — extend regexes if Claude / GPT / Llama produce a shape not covered
- [ ] Cost-estimate a 1000-sample × 3-judge real run before kicking it off
- [ ] Decide whether to log raw model responses alongside parsed scores for audit (currently only the parsed `score` + `rationale` are persisted)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: swarm/judges/llm_call.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    `sleep` is injectable so tests can run without wall-clock waits.
+    Retries every Exception; if you want narrower handling, wrap before
+    calling.
+    """
+    last: Optional[BaseException] = None
+    for attempt in range(max_retries + 1):
+        try:
+            return fn()
+        except Exception as err:
```
Comment (rsavitt) [inline swarm/judges/llm_call.py:293]: [ISSUE] `call_with_retries` catches `except Exception`, so terminal failures — `ImportError` from a missing SDK, `RuntimeError` from a missing API key, 4xx auth/quota errors — all consume the full retry budget instead of failing fast. Misconfiguration takes 4× longer to surface, and the warning log buries the real cause behind transient-looking timing. Narrow this to network/5xx/timeout exceptions, or accept a retryable-exception predicate.

File: swarm/judges/judge.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            interaction_id=view.interaction_id,
+            judge_name=self.name,
+            rubric_version=RUBRIC_VERSION,
+            score=score,
+            rationale=rationale,
+        )
+
+
+JSON_FENCE_RE = re.compile(r"```(?:json)?\s*(\{.*?\})\s*```", re.DOTALL)
+JSON_OBJECT_RE = re.compile(r"(\{[^{}]*\"score\"[^{}]*\})", re.DOTALL)
```
Comment (rsavitt) [inline swarm/judges/judge.py]: [ISSUE] JSON_OBJECT_RE = r"(\{[^{}]*\"score\"[^{}]*\})" rejects any score object that contains nested JSON values — e.g. {"score": 0.5, "evidence": {"agent_type": "honest"}} or a rationale containing escaped braces. The [^{}]* class hard-stops at the first inner brace, so the regex won't match and _extract_score falls through to ValueError on a common LLM output shape, silently dropping the row instead of recording its score. Either allow nested braces (with balanced parsing) or constrain the rubric to flat objects and enforce that in the prompt.

File: swarm/judges/judge.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            "rate the interaction strictly per its instructions. Return ONLY "
+            "a JSON object matching the rubric's output format.\n\n"
+            "===== RUBRIC =====\n"
+            f"{rubric}\n"
+            "===== INTERACTION =====\n"
+            f"{json.dumps(payload, indent=2)}\n"
+            "===== OUTPUT =====\n"
+        )
+
+    def _dispatch(self, prompt: str) -> LLMCallResult:
```
Comment (rsavitt) [inline swarm/judges/judge.py:271]: [ISSUE] LLMJudge's contract — three independent judges scoring the same view — has no test that the dispatch actually reaches three different providers at HEAD. The 15 LLMJudge tests use the `caller` injection to bypass `_dispatch`, so a misrouted provider string (e.g. typoing `"openrouter"` so it falls through to `raise ValueError`) only surfaces in a live calibration run. Add a parametrized test that asserts each `provider` value (`anthropic`, `openai`, `openrouter`, `groq`, `together`, `deepseek`, `ollama`) routes to the expected `call_*` function by patching the helpers — keeps the dispatch table honest.

**Review by rsavitt — COMMENTED** [review]
**Verdict**: REQUEST_CHANGES — quality-gate evidence missing at HEAD, plus three reliability gaps in the new judge pipeline.
**Confidence**: 2/5

**Evidence**:
- Passed: Vercel preview (`Vercel – swarm`, `Vercel Preview Comments`).
- Missing/failing: lint, type-check, test 3.10/3.11/3.12 (compatibility/full), Memory Tests, CodeQL, quality-gate — none ran on `e216382d`. The PR body states "32 tests pass; ruff + mypy clean" but there is no machine evidence at this SHA. Under merge-gate policy ("APPROVE only if risk is low AND evidence is sufficient"), this fails the evidence bar even before considering the inline findings.
- PR `baseRefName` is `main`, but the body says "Stacked on #485 (arm B). Base on this PR is `feat/calibration-arm-b`." The diff re-introduces 7 files that #485 owns (`swarm/judges/{__init__.py,rubric_v1.md,sampler.py,views.py}`, `swarm/calibration/{__init__.py,fidelity.py}`, `experiments/calibration_fidelity.py`). The stack is either inverted or out of sync — merge ordering with #485/#487 needs to be resolved before this can land cleanly.

**Findings** (mirrored as inline comments):
- [ISSUE] swarm/judges/llm_call.py:222 — `call_with_retries` catches every `Exception`, so `ImportError` (missing SDK), `RuntimeError` (missing API key), and 4xx auth/quota errors all consume the full retry budget instead of failing fast. Narrow the catch or accept a retryable-exception predicate.
- [ISSUE] swarm/judges/judge.py:141 — `JSON_OBJECT_RE` rejects score objects containing nested JSON values; common LLM outputs (`{"score": 0.5, "evidence": {...}}`) silently fall through to `ValueError` and the row is dropped.
- [ISSUE] swarm/judges/judge.py:234 — `_dispatch` provider table is untested at HEAD (all LLMJudge tests use the `caller` injection). A typoed provider string only surfaces in a live calibration run; add a parametrized routing test.

Path to APPROVE: rebase onto a SHA that runs the standard CI matrix (or trigger it manually), tighten retry semantics, broaden the score-extraction regex, and confirm stack ordering with #485/#487.

**Review by rsavitt — COMMENTED** [review]
**Verdict**: REQUEST_CHANGES
**Confidence**: 3/5

Code-level concerns from the prior review at SHA `e216382d` are all addressed at the new SHA `656718e6` — but the merge-gate evidence bar is still unmet at this HEAD.

**Evidence**:
- Passed at HEAD: CodeQL (python + js), Vercel preview.
- Missing at HEAD: lint, type-check, test (3.10/3.11/3.12), Memory Tests, render-verify, kb-graph-check, quality-gate. Body claims "32 tests pass; ruff + mypy clean" — no machine evidence on this commit. Sibling refactor #489 ran the full matrix at the same time window from `main`, so the matrix is operational; it's specifically not firing for this branch (consistent with the body's "Stacked on #485" / `baseRefName=main` mismatch).

**Resolved from prior review (good)**:
- ~~`swarm/judges/llm_call.py` retry-on-every-Exception~~ — `default_retryable` + `_RETRYABLE_EXC_NAMES` now fail fast on `ImportError` / 4xx / missing-key.
- ~~`swarm/judges/judge.py` regex `JSON_OBJECT_RE` rejects nested objects~~ — replaced with `_iter_brace_objects` (string-aware balanced-brace parser).
- ~~`_dispatch` provider table untested~~ — `test_dispatch_routes_to_expected_provider` parametrizes each provider string and verifies the real `call_*` is invoked.

**Path to APPROVE**:
- Get the full CI matrix to actually run at HEAD (likely needs the stack vs. `main` discrepancy resolved — either rebase on `main` once #485 lands, or land #485 first and re-trigger).
- Confirm overlap with #487 (`swarm/judges/judge.py` has divergent versions on both PRs) — whichever merges last must be the canonical one.

**Comment by rsavitt** [comment]
Rebased onto `origin/main` (4d78e8c7) to clear the conflict that was suppressing the `pull_request` workflow synthesis. Dropped the now-redundant arm-A/arm-B commits — those are on main as PRs #484 / #485. Branch is now four commits: (1) `feat: wire LLMJudge`, (2) `fix: address review — endpoint defaults, fail-fast retries, nested-JSON, dispatch test`, (3) `fix: protocol docstring body`, (4) `fix: address PR #488 review nits — stale docs, regex, edge cases`. CHANGELOG has a single consolidated wiring entry at the top of `### Added` and the leftover "typed stub that raises NotImplementedError" arm-B clause is past-tensed. Full CI matrix is running on `4d78e8c7` (`MERGEABLE`, no longer `CONFLICTING`); 50/50 judge + llm_call tests pass locally, ruff + mypy clean.

**Review by rsavitt — COMMENTED** [review]
**Verdict**: APPROVE — code-level concerns from prior SHAs all addressed, full merge-gate evidence now present at HEAD (4d78e8c7).
**Confidence**: 5/5.

(Posting as a comment rather than approval since GH disallows self-approve.)

**Evidence**:
- Passed at `4d78e8c7`: lint, type-check, test (3.10/3.11/3.12), Memory Tests 3.10/3.11/3.12 + baseline, kb-graph-check, render-verify, quality-gate, CodeQL (python + js), Vercel preview.

**Substantive concerns resolved on the branch:**
- `call_with_retries` no longer eats terminal failures — default classifier retries 5xx/429/timeout, fails fast on 4xx/missing-SDK/missing-key.
- JSON parsing now uses brace-aware `_iter_brace_objects` instead of `JSON_OBJECT_RE`, so nested objects in `score` payloads parse cleanly.
- Provider-dispatch test (`test_dispatch_routes_to_expected_provider`) covers the three real provider strings — anthropic/openai/ollama — so a typo'd `JUDGE_SPECS.provider` would be caught.
- Built-in endpoint defaults for non-OpenAI OpenAI-compatible providers (OpenRouter / Groq / Together / DeepSeek) prevent silent mis-routing to `api.openai.com`.
- Resolved `provider`+`model` per judge is now persisted to `config.json` so two runs both labelled `gpt4o_mini` but pointed at different models via `JUDGE_MODEL_GPT4O_MINI` are distinguishable from artifacts.

**Nits resolved at HEAD (this PR):**
- `JSON_FENCE_RE` non-greedy.
- `call_openai_compatible` docstring no longer cites `ANTHROPIC_API_KEY`.
- Ollama URL `.rstrip('/')` so trailing-slash `base_url` doesn't produce `//api/chat`.
- `calibration_judge` module docstring + `--judges` help describe real credential requirements.
- `calibration_fidelity` no longer emits a trailing comma in the `gap` CSV cell.
- CHANGELOG has a single consolidated wiring entry; arm-B entry's "typed stub that raises NotImplementedError" clause is past-tensed.
- Module docstring on `judge.py` updated to describe `LLMJudge` as it actually is now.

**Rebase note:** branch was rebased onto current main and force-pushed to clear a merge conflict that was suppressing `pull_request` workflow synthesis (which is why prior verdicts cited missing CI matrix evidence). Dropped two now-redundant commits whose contents are on main as #484 (arm A) / #485 (arm B); kept only the four LLM-wiring commits.
